### PR TITLE
Real to complex

### PR DESCRIPTION
**Pull request by Lagrang3**
In this PR we introduce the real-to-complex (and the complementary complex-to-real) transforms. The design follows the discussion started in the issue #14.

### CODE REVIEW

**Comment by ckormanyos** [comment]
Hi @Lagrang3 this looks like good progress.

It's my own weakness, but I was trying to create a simple multiplication example using R2C (multiply via convlution in the transform space) for two real sequences a and b. But I was unable to fully comprehend the syntax needed.

Do you have/could you create a multiplication example that we might build/test as an example in CI?

**Comment by Lagrang3** [comment]
> Hi @Lagrang3 this looks like good progress.
> 
> It's my own weakness, but I was trying to create a simple multiplication example using R2C (multiply via convlution in the transform space) for two real sequences a and b. But I was unable to fully comprehend the syntax needed.
> 
> Do you have/could you create a multiplication example that we might build/test as an example in CI?

The example you are proposing is a nice application and test.
I'm still not happy with this Real to Complex interface, because of the asymmetry between the real and complex types and the size of the input and output. I will take the opportunity, while working on the example, to think of a good interface. 
Thanks.

**Comment by ckormanyos** [comment]
> think of a good interface

Right on! I like the way you approach this detailed interface.

I will also take more time to use your branch(es) and this draft and think if I can find any sensible ideas along the lines you suggest.

**Comment by Lagrang3** [comment]
> > think of a good interface
> 
> Right on! I like the way you approach this detailed interface.
> 
> I will also take more time to use your branch(es) and this draft and think if I can find any sensible ideas along the lines you suggest.

Notice that up to [this commit](https://github.com/BoostGSoC21/math/pull/15/commits/6e706932d8aef10948a2d769590b0c82d1a94209) we have two types of FFT plans: one for purely algebraic types (one possible application is NTT, fast integer multiplication, etc) and another for complex numbers. In the complex number plan I added real transforms. But like I said before (#14) I think is better if we have a separate plan real-to-complex/halfcomplex. This idea is not implemented yet.

Regarding the real interface, there is one observation I want to share with you. The input is an array A[k] of N real numbers and this will naively produce an array of N complex numbers with a conjugate symmetry B[k] = (conjugate B)[N-K]. Half of this complex numbers encode the information, the other half is just complex conjugate. Therefore interfaces like python's numpy and FFTW tend to output only those complex numbers which are relevant; ie. there will be M=floor(N/2)+1 of those. Notice that the reverse operation needs to know N before-hand because floor(N/2)+1 forgets the parity of N.
For example take N=5, in the output M=3 complex numbers are sufficient to recover all the information. But if we are given M=3 complex numbers, there is no way of knowing if they come from N=5 real or N=4.
The number of degrees of freedom in the output is still equal N, because the first complex has 0 imaginary part, as well as the last in the case of N even. Hence, the gsl interface likes to produce N output real numbers in a fashion that they call halfcomplex that encode the unique complex components it would produce the FFT.
For instance, for N=5 they halfcomplex sequence represents the complex outputs in the following way
```
complex[0].real = halfcomplex[0]
complex[0].imag = 0
complex[1].real = halfcomplex[1]
complex[1].imag = halfcomplex[2]
complex[2].real = halfcomplex[3]
complex[2].imag = halfcomplex[4]
// M = floor(5/2) +1 = 3, up to here, the rest are redundant complex numbers
complex[3].real = halfcomplex[3]
complex[3].imag = -halfcomplex[4]
complex[4].real = halfcomplex[1]
complex[4].imag = -halfcomplex[2]
```
for N = 6 instead, we have
```
complex[0].real = halfcomplex[0]
complex[0].imag = 0
complex[1].real = halfcomplex[1]
complex[1].imag = halfcomplex[2]
complex[2].real = halfcomplex[3]
complex[2].imag = halfcomplex[4]
complex[3].real = halfcomplex[5]
complex[3].imag = 0
// M = floor(6/2) +1 = 4, up to here, the rest are redundant complex numbers
complex[4].real = halfcomplex[3]
complex[4].imag = -halfcomplex[4]
complex[5].real = halfcomplex[1]
complex[5].imag = -halfcomplex[2]
```
They also provide a routine to convert (that is unpack in their language) the halfcomplex into N complex numbers.
Initially I preferred the FFTW and python interface, but it will be easier to stay always real and transform N real to N halfcomplex like GSL does, and providing tools on-demand to decode the halfcomplex into complex.

**Comment by Lagrang3** [comment]
I've uploaded an example of polynomial multiplication (real sequences convolution) using the real to complex interface. Details are yet to polish.

**Comment by ckormanyos** [comment]
> uploaded an example of polynomial multiplication

Thank you Eduardo.

@Lagrang3 

>  easier to stay always real and transform N real to N half complex

I have also always preferred this approach. In fact, I have a similar example posted in my [wide-decimal](https://github.com/ckormanyos/wide-decimal) project. In that project, the multiplication of the half-complex arrays in the transform space is [here](https://github.com/ckormanyos/wide-decimal/blob/593852f8c2c268cd4c5afd24d857a1192557c1d9/math/wide_decimal/decwide_t_detail_ops.h#L515), in case that helps us in the future. I believe my storage convention is the same one used by FFTW's _rfft_.

**Comment by cosurgi** [comment]
@Lagrang3 I pushed two commits here, expanding the example and testing to more types.

**Comment by cosurgi** [comment]
I guess you are right now working on `fft::bsl_rfft`.

**Comment by Lagrang3** [comment]
> I guess you are right now working on `fft::bsl_rfft`.

That's right. 
But still I wonder if we should keep the halfcomplex format that I've choosen, which is the GSL style (see the description [above](https://github.com/BoostGSoC21/math/pull/15#issuecomment-890859962)). There is another possibility, which is the FFTW halfcomplex format. 

In the FFTW format the imaginary part of the positive frequency amplitudes are saved starting from the end of the halfcomplex array coming backwards, like iterating with negative indexes. For example for N=5
```
complex[0].real = halfcomplex[0]
complex[0].imag = 0
complex[1].real = halfcomplex[1]
complex[1].imag = halfcomplex[5-1]
complex[2].real = halfcomplex[2]
complex[2].imag = halfcomplex[5-2]
// M = floor(5/2) +1 = 3, up to here, the rest are redundant complex numbers
complex[3].real = halfcomplex[5-3]
complex[3].imag = -halfcomplex[3]
complex[4].real = halfcomplex[5-4]
complex[4].imag = -halfcomplex[4]
```
for N = 6 instead, we have
```
complex[0].real = halfcomplex[0]
complex[0].imag = 0
complex[1].real = halfcomplex[1]
complex[1].imag = halfcomplex[6-1]
complex[2].real = halfcomplex[2]
complex[2].imag = halfcomplex[6-2]
complex[3].real = halfcomplex[3]
complex[3].imag = 0
// M = floor(6/2) +1 = 4, up to here, the rest are redundant complex numbers
complex[4].real = halfcomplex[6-4]
complex[4].imag = -halfcomplex[4]
complex[5].real = halfcomplex[6-5]
complex[5].imag = -halfcomplex[5]
```

**Comment by cosurgi** [comment]
This looks like some kind of container view. A different view on the same data. Similar to viewing on a slice of a container (e.g in Boost.Ublas). We could just keep in mind to not block such approach in the future.

**Comment by Lagrang3** [comment]
> This looks like some kind of container view. A different view on the same data. Similar to viewing on a slice of a container (e.g in Boost.Ublas). We could just keep in mind to not block such approach in the future.

Good observation! I was getting into troubles because of the different ways of representing the same object: the halfcomplex array. The halfcomplex then could be represented as a container with N space requirement (instead of 2N for N complex numbers), etc, with straightforward user interface access to the real or imaginary part of the individual complex numbers.
We could eventually have the `real_to_halfcomplex` return this new kind of halfcomplex container.
In order to make this compatible with GSL and FFTW we could add a halfcomplex factory specific for each of these backends; e.g. `halfcomplex<double> gsl_create_halfcomplex(double* first, double* last);`

File: include/boost/math/fft/multiprecision_complex.hpp [hunk 1]
```diff
@@ -12,9 +12,9 @@
 //                                                         it has nothing to do with boost::is_complex<T> which recognizes only std::complex<T>
 //
 // TODO: this file should be split apart and each specializetion should go to the end of the relevant file in Boost.Multiprecision
-//   → specialization for float128      → complex128    goes to <boost/multiprecision/complex128.hpp>
-//   → specialization for cpp_bin_float → cpp_complex   goes to <boost/multiprecision/cpp_complex.hpp>
-//   → specialization for mpfr_float    → mpc_complex   goes to <boost/multiprecision/mpc.hpp>
+//   --> specialization for float128      --> complex128    goes to <boost/multiprecision/complex128.hpp>
```
Comment (Lagrang3) [inline include/boost/math/fft/multiprecision_complex.hpp:15]: in fact. I noticed @cosurgi often uses non-ascii characters.

**Comment by ckormanyos** [comment]
> uses non-ascii characters

It does not really matter these days. I was hard-boiled last century (last millenium), and trained to use only ascii chars in code. Today's compilers handle wide characters. I don't even know if Boost has some kind of ascii rule. But I would prefer old-school ascii if it's not too much of a hassle.

**Comment by ckormanyos** [comment]
Hi @Lagrang3 while we are on such a roll, should we push all this to develop (of the fork).

CI is looking good, so from the administrative side all OK. But i know you are still working a bit on the content of `rfft`. Should we push this to develop just to keep up to date (even if a few details are still ongoing)? Or would you prefer to wait?

**Comment by Lagrang3** [comment]
go ahead with the merge. I can create a new PR dedicated to the boost rfft backend.

**Comment by ckormanyos** [comment]
> go ahead with the merge.

Done. Many thanks, Eduardo and Janek.

> I can create a new PR dedicated to the boost rfft backend.

Perfect. Thanks.

@Lagrang3 and @cosurgi

File: include/boost/math/fft/multiprecision_complex.hpp (hunk 1 above)
Comment (cosurgi) [reply include/boost/math/fft/multiprecision_complex.hpp:15]: Heh, yeah. Several years ago I figured how to use compose key :) And wrote into `~/.XCompose` config with what I like to write.

**Comment by cosurgi** [comment]
Great work! Thank you very much for re-checking this. I will limit a bit my usage of UTF8 chars :)

**Pull request merged**